### PR TITLE
Clean-up Python #4815 - P8.12: ML module

### PR DESCRIPTION
**Pull request by IchiruTake**
This is the sub-PR breaking from the large PR 4815: https://github.com/rdkit/rdkit/pull/4815

--------------
Note: This is my final PR from 4815. There are still some further work to smooth out and modernize the Python codebase in RDKit such as f-string, logic-handling, and algorithm. However, there are some notes that I want to cover, which probably in the near future due to the close schedules of my internship after Lunar New Year (I can still have free time in February but not March)

- [F-String]: However, there are several concerns that push extra work to clarify the underlying behavior of f-string in Python, especially those working with RDKit's database. However, some string code in old Python (< 3.5 or 2) cannot be re-formatting into f-string without checking. I am so sorry that I cannot fulfill this responsibility 100 %. 
- [Optimization Note]: There are several concerns about the optimization such as if - and/or (#4914 ), and probably other places. As mentioned, every work and optimization are naive and fully-blind (or partially-blind if correctly stated). I hope that expect that every significant and non-significant changes I has made can all be reported in text and code to track down.
- [Leftover]: Due to the basic need of RDKit's user, I decided to drop cleanup in `sping` module which has been mentioned in PR #4815 . I hope that after my internship and final project for undergraduate, I can come back and push up all of my leftover work, including cleanup for `sping` module, modernize all remaining f-string, provide datatype typing, and probably optimize the algorithm (in C++ and Python as well).

Thank you for all authors and contributors of the RDKit has supported me from my first hand-on experience on contributing the first giant PR.

Final Word: I can still have free time in February but not March so feel free to tag me in my PR-series that has trouble so that I can fix it immediately.

### CODE REVIEW

**Comment by IchiruTake** [comment]
- [Note]: Weird string such as `'% 4.2f' % (float)` = `' float'` but `'% s' % string` = `'string'`, but `'%  4.2f' % (float)` = `' float'`. These are some basic examples of weird string that I needs some time to consume. It is hard to track down those issue. (Test is conducted in Python 3.8.10 in GDB and Python 3.9 (local)). What do you think about this issue, @gedeck ?
- Also, if some module received less attention and less need of string manipulation, I usually ignored it.

File: rdkit/ML/AnalyzeComposite.py [hunk 1]
```diff
@@ -113,16 +112,18 @@ def ErrorStats(conn, where, enrich=1):
   if not nPts:
     sys.stderr.write('no runs found\n')
     return None
-  overall = numpy.zeros(nPts, numpy.float)
-  overallEnrich = numpy.zeros(nPts, numpy.float)
+  
+  overall = numpy.zeros(nPts, dtype='float')
```
Comment (bp-kelley) [inline rdkit/ML/AnalyzeComposite.py:116]: Here the dtype can just be (float is the default):

```
  overall = numpy.zeros(nPts)
```

or

```
 overall = numpy.zeros(nPts, dtype=float)
```

We only need to use strings representations for more complicated types

File: rdkit/ML/ScreenComposite.py [hunk 2]
```diff
@@ -706,25 +698,23 @@ def ScreenFromDetails(models, details, callback=None, setup=None, appendExamples
                                 pickleClass=DataStructs.ExplicitBitVect)
     else:
       data = details.GetDataSet()
-  if details.threshold > 0.0:
-    details.partialVote = 1
-  else:
-    details.partialVote = 0
+  
+  details.partialVote = int(bool(details.threshold > 0.0))
```
Comment (bp-kelley) [inline rdkit/ML/ScreenComposite.py:702]: I'm not sure I like this reformatting.

The "pythonic" way is
```
details.partialVote = 1 if details.threshold > 0.0 else 0
```

File: rdkit/ML/ScreenComposite.py [hunk 3]
```diff
@@ -706,25 +698,23 @@ def ScreenFromDetails(models, details, callback=None, setup=None, appendExamples
...
   if setup is not None:
     setup(nModels * data.GetNPts())
 
-  nGood = numpy.zeros(nModels, numpy.float)
-  nBad = numpy.zeros(nModels, numpy.float)
-  nSkip = numpy.zeros(nModels, numpy.float)
-  confGood = numpy.zeros(nModels, numpy.float)
-  confBad = numpy.zeros(nModels, numpy.float)
-  confSkip = numpy.zeros(nModels, numpy.float)
+  nGood = numpy.zeros(nModels, dtype='float')
```
Comment (bp-kelley) [inline rdkit/ML/ScreenComposite.py:712]: dtype=float

File: rdkit/ML/ScreenComposite.py [hunk 4]
```diff
@@ -1397,14 +1386,14 @@ def ParseArgs(details):
       if len(details.screenVoteTol) > 1:
         message('\n-----*****-----*****-----*****-----*****-----*****-----*****-----\n')
         message('Tolerance: %f' % tol)
-      nGood = numpy.zeros(nModels, numpy.float)
-      nBad = numpy.zeros(nModels, numpy.float)
-      nSkip = numpy.zeros(nModels, numpy.float)
-      confGood = numpy.zeros(nModels, numpy.float)
-      confBad = numpy.zeros(nModels, numpy.float)
-      confSkip = numpy.zeros(nModels, numpy.float)
+      nGood = numpy.zeros(nModels, dtype='float')
```
Comment (bp-kelley) [inline rdkit/ML/ScreenComposite.py:1389]: dtype=float

**Review by bp-kelley — CHANGES_REQUESTED** [review]
Thanks for working on this.

File: rdkit/ML/BuildComposite.py [hunk 5]
```diff
@@ -522,13 +521,13 @@ def RunOnData(details, data, progressCallback=None, saveIt=1, setDescNames=0):
       silent=not _verbose)
 
   elif details.useSigTrees:
-    from rdkit.ML.DecTree import CrossValidate
-    from rdkit.ML.DecTree import BuildSigTree
+    from rdkit.ML.DecTree import BuildSigTree, CrossValidate
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/ML/BuildComposite.py:524]: I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/ML/ScreenComposite.py (hunk 2 above)
Comment (IchiruTake) [reply rdkit/ML/ScreenComposite.py:702]: I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/ML/ScreenComposite.py [hunk 6]
```diff
@@ -776,32 +766,30 @@ def ScreenFromDetails(models, details, callback=None, setup=None, appendExamples
...
+    cb = None
     if callback:
       cb = lambda x, y=callback, z=i * data.GetNPts(): y(x + z)
-    else:
-      cb = None
-    if not hasattr(details, 'errorEstimate') or not details.errorEstimate:
-      errorEstimate = 0
-    else:
-      errorEstimate = 1
+    errorEstimate = int(not bool(not hasattr(details, 'errorEstimate') or not details.errorEstimate))
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/ML/ScreenComposite.py:775]: I got your idea! I would fix it in the next patch. Thank you for your review.

File: rdkit/ML/ScreenComposite.py [hunk 7]
```diff
@@ -426,7 +419,7 @@ def ShowVoteResults(indices, data, composite, nResultCodes, threshold, verbose=1
     print('average correct confidence:   % 6.4f' % avgGood)
     print('average incorrect confidence: % 6.4f' % avgBad)
 
-  voteTab = numpy.zeros((nResultCodes, nResultCodes), numpy.int)
+  voteTab = numpy.zeros((nResultCodes, nResultCodes), dtype='int')
```
[the remark this answers is not in the document]
Comment (IchiruTake) [reply rdkit/ML/ScreenComposite.py]: Yes, I have fixed this

**Comment by IchiruTake** [comment]
@bp-kelley I have changed as requested, can you take a look on this PR to review? Thank you very much

**Pull request closed without merging**